app/containers/App/appSlice.js

```diff
@@ -0,0 +1,39 @@
+import { createSlice } from '@reduxjs/toolkit';
+import appReducer from './reducer';
+
+// The initial state of the App
+export const initialState = {
+  loading: false,
+  error: false,
+  currentUser: false,
+  userData: {
+    repositories: false,
+  },
+};
+
+/* eslint-disable default-case, no-param-reassign */
```

Comment:
Could you talk more about the way the param reassignment works here? I'm not sure how I feel about turning off a lint rule to accommodate the way a package works. I also think that as I'm reading this, it comes off as if we're mutating state here by reassigning the state params. This might just be me not fully understanding the way this works under the hood.

Reply:
Sure.

RTK specifically uses the [Immer library](https://immerjs.github.io/immer/docs/introduction) library inside.  Immer lets you write "mutating" code, but it wraps the original value in an ES6 Proxy and tracks the modifications, and then safely returns a correctly immutably updated value (as if you'd done all the spreads and stuff by hand).

`react-boilerplate` is already using Immer in its reducers, and in fact I copied the `eslint-disable` line from another file, like the equivalent `App/reducer.js`:

https://github.com/react-boilerplate/react-boilerplate/blob/27538f1595620b235eae2ae4251305da82dc386a/app/containers/App/reducer.js#L23-L30

So, in that sense there's nothing "new" here for `react-boilerplate`, this is already stuff you've got in the codebase.  The only changes are A) the additional `@reduxjs/toolkit` package, B) the `createSlice()` function auto-generating the action creators and action types, and C) consolidating the resulting logic into one file instead of several.

Reply:
Yes, we had to disable these rules for our migration to `immer`. You should only need `/* eslint-disable no-param-reassign */` here though as `default-case` is only for switches.

Reply:
Word, thanks for the explanations. I like to make my day-to-day life difficult by not really ever using immutability libs, so this is helpful.

Reply:
lol yes @gretzky i'm with you—super clever implementation but personally not a fan of immer...setting properties using dot notation just _feels_ wrong. long live explicit immutability 😄

Reply:
The good news is that using RTK and Immer doesn't _require_ that you actually use "mutative" logic - you can still return manually written immutable updates.

That said, frankly hand-written immutable update code is overly verbose, hard to read, and very prone to mistakes.  I understand the concerns about this "feeling wrong", but using Immer is actually much safer _and_ easier to read.

Reply:
I don't think there's anything inherently wrong with dot notation for writing immutable updates but hiding `immer` can IMO make things confusing for new users. They might not realize that they're not **actually** mutating the state.

I was previously against this when [someone suggested switching to `redux-immer`](https://github.com/react-boilerplate/react-boilerplate/issues/2775#issuecomment-542213763). I want to make sure that we make it as clear as possible (via docs and/or comments) that there's some work in the bg making it possible to write immutable logic with dot notation.

Reply:
Yeah, the teaching aspect is definitely the hardest part here.